content/posts/redis-modules-into-arm-land-part1.md

```diff
@@ -0,0 +1,275 @@
+
+If you'll follow through, you'll end up with a fully functional ARM build laboratory.
+
+## Inside RedisEdge
+
+![redisedge1](/redis-modules-arm-1.png)
+
+But first, let's take a look at RedisEdge.
+RedisEdge is not a Redis module, but an aggregate of three Redis modules: RedisGears, RedisAI, and RedisTimeSeries.
+It is distributed as a Docker image, which is based on Redis Server 5.0. Thus, one can simply pull the image, run it, and start issuing Redis commands, load models into RedisAI, and execute Python gears scripts on RedisGears.
+Although one can easily get Docker out of the equation (by installing a Redis server and copying Redis Modules files), we'll see that Docker does actually provide significant added value, and it is worthwhile keeping it around.
+
+![redis-edge-1](/redis-modules-arm-2.png)
+
+*Inside RedisEdge: modules structure*
+
+Let's now take a look at each of the components of RedisEdge, to try to figure out what would it take to have them ported to ARM.
+First, **[RedisTimeSeries](https://github.com/RedisTimeSeries/RedisTimeSeries)**. 
+It's a simple C library, built with make. Not even a configure script. No problems there.
+Next is [**RedisGears**](https://github.com/RedisGears/RedisGears). It's too a C library built with plain make, but it does use an embedded Python 3.7 interpreter, which is built from source. This requires running automake to generate a platform-specific makefile.
+Finally, [**RedisAI**](https://github.com/RedisAI/RedisAI). It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like [TensorFlow](https://github.com/tensorflow/tensorflow), [PyTorch](https://github.com/pytorch/pytorch), and [ONNXRuntime](https://github.com/microsoft/onnxruntime), in their C library form (most users typically use them in Python), with PyTorch and ONNXRuntime not officially supporting ARM.
+So the build requirements, as it seems, have deteriorated quickly. It went from building an innocent C library to compiling massive source bases with convoluted build systems.
+
+In the next sections, we will fit each components with its proper build method.
+
+## Building for ARM
+
+At this point, we'll pause and take stock of what is required to build software for ARM.
+The obvious way is to use an ARM-based device, like Raspberry Pi. Once set up, you'll be able to build and test in the most natural manner. 
+
+The testing medium is important: even though you can build almost every ARM software without a physical ARM device, there is no reliable way of testing the outcome in a reliable way without such a device, especially with non-standard devices. Therefore, as much as virtualized/emulated/containerized test may be useful, you should always test your software on its designated target device.

```

Comment:
Change to: "The testing medium is important: even though you can build almost every ARM software without a physical ARM device, there is no way to reliably test the outcome without such a device, especially with non-standard devices. Therefore, while a virtualized/emulated/containerized test may be useful, you should always test your software on its designated target device."